### PR TITLE
pfblockerng: opt-in retention buffer on scheduled log reset (#416)

### LINKED ISSUE #416: ADR-30 amendment: opt-in retention buffer on scheduled log reset (for log-shipper users)
## Motivation

The ADR-30 scheduled reset (`pfb_log_reset()`) truncates a log to **empty** (`cp /dev/null`) at each calendar boundary. For the common case — users reading logs in the pfBlockerNG GUI (Reports/Alerts) — a clean "fresh log per period" boundary is exactly right.

But for the **log-shipper audience** (remote syslog / a shipper keyed on `(inode, offset)` — the same niche that motivated the inode-preserving truncation in #264), the full empty can discard the **unshipped tail**: lines written after the shipper's last read but before the boundary truncation are lost. The line-cap path (`pfb_log_mgmt`, `tail -n N`) doesn't have this problem (it drops the *oldest* lines, which have had the most time to ship); only the boundary reset zeroes the newest content.

## Proposal

Add an **opt-in, per-log retention buffer** to the scheduled reset:

- New per-log setting `log_reset_keep_<logtype>` (config gateway / ADR-29 registry), **default `0`**.
- `0` ⇒ current behavior exactly (`cp /dev/null`, full empty) — no change for anyone not opting in.
- `> 0` (= K) ⇒ on reset, keep the **last K lines** via the existing inode-preserving `tail -n K` mechanism (`pfb_log_mgmt` already does this), instead of zeroing.
- Marker/period bookkeeping unchanged (still records the new period key once per boundary).

## Why a line count, not a time window

The request was framed as "keep the last few minutes/hours." A true time window would need to parse a timestamp from every line, and the 10 log types have heterogeneous formats (some lines aren't cleanly timestamped) — fragile and format-coupled. A line count is a robust proxy that reuses proven, format-agnostic machinery; the help text should be upfront that it's a line cushion, not a literal time guarantee.

## Rejected alternatives

- **Rotate to a `.1` sidecar instead of emptying** — breaks the inode-preserving invariant (#264); a shipper keyed on `(inode, offset)` would see a rotation and re-ship the whole file. This is the exact problem ADR-30 avoids.
- **Time-window retention (keep lines newer than T)** — per-log-format timestamp parsing; deferred unless a concrete need appears.

## Scope / priority

Niche (only log-shipper setups), low priority, fully backward-compatible (default `0`). Captured here as an ADR-30 amendment so it's tracked for whenever a shipper user needs it.

## Notes

Surfaced during #414 / PR #415 (the smoke-trigger fix). No relation to that fix beyond touching the same `pfb_log_reset()` area.

### PR DESCRIPTION
**Pull request by andrebrait**
## Summary

Implements the ADR-30 amendment from issue #416: an **opt-in, per-log retention buffer** on the scheduled log reset, for operators shipping pfBlockerNG logs off-box.

## Motivation

The ADR-30 scheduled reset truncates a due log to **empty** at each calendar boundary. For remote log shippers (keyed on `(inode, offset)` — the #264/#280 audience), the full empty can discard the **unshipped tail**: lines written after the shipper's last read but before the boundary. The line-cap trim doesn't have this problem (it drops the *oldest* lines); only the reset zeroes the newest content.

## Change

A new per-log `log_reset_keep_<type>` setting (numeric, **default `0`**), mirroring `log_rotate_<type>` / `log_max_<type>` exactly:

- **`0`** (default) → full empty (`cp /dev/null`) — **byte-for-byte the current behavior**, no change for anyone not opting in.
- **`> 0` (= K)** → keep the **last K lines in place** via the same inode-preserving `tail -n K → temp → cat` mechanism as `pfb_log_mgmt()`, with per-exec exit-code gating (a failed `tail`/`cat` never blanks the live log) and the same chroot-path + `chown unbound` handling for the DNS python logs.
- Values that cast to `< 1` are treated as full empty. Marker/period bookkeeping is **identical** for both paths — idempotency and per-log independence unchanged.

**Line count, not time window:** the 10 log types have heterogeneous formats (some lines aren't cleanly timestamped), so a true time window would need fragile per-format parsing. A line count is a robust, format-agnostic proxy; the UI help text says so. (Rejected: a `.1` sidecar — it would re-introduce the #264/#280 inode-churn the design avoids.)

## Commits

1. `register per-log log_reset_keep config field + UI` — registry loop entry + General-settings field + gateway/rollback/sniff tests (behavior-preserving).
2. `keep last N lines on scheduled log reset when configured` — the `pfb_log_reset()` branch + `LogRotateResetTest` coverage.
3. `amend ADR-30 with the opt-in log-reset retention buffer` — ADR-30 §8.
4. `live-VM coverage for log-reset retention buffer` — smoke cases.

## Validation

- Off-box: `phpunit` 1110, `pytest` 1904, `phpcs`/`phpstan`/`ruff`/`markdownlint` all green.
- **Branch coverage** (`LogRotateResetTest.php`): `K=0` empties; `K>0` keeps exactly the last K lines (content + **inode** preserved); fewer-than-K retains the whole file; plain and dns-log paths. The `K>0` assertion was confirmed to fail on the pre-amendment code — proving it's a real branch.
- Config round-trip + rollback invariants for all 10 new keys.
- **Live-VM** (`tests/smoke/test_log_rotate.py`): two new `K>0` cases (plain + chrooted dns log, asserting last-K content, inode, and `unbound` ownership), paired with the existing `K=0` cases. Dispatched on this branch.

Fixes #416

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01WWUuidswrN5TNyfBHwHqgt)_

## Release Notes

* **New Features**
  * Added per-log “Keep lines” settings in General Log Settings to optionally retain the last *N* lines during scheduled log resets.
  * Supports full clearing when set to `0`, and tail retention (including fewer-than-*N* lines) when set above `0`.

* **Documentation**
  * Updated the scheduled log reset design record to cover the new “keep lines” amendment behavior and edge cases.

* **Tests**
  * Expanded unit and smoke testing for configuration validation, round-trip/rollback defaults, and reset file/marker behavior (including DNS/chroot handling and ownership).

### CODE REVIEW

File: tests/smoke/test_log_rotate.py [hunk 1]
```diff
@@ -571,3 +632,185 @@ def test_all_off_no_log_emptied(deployed_vm: SmokeVM) -> None:
...
+    assert marker_after.get("ip_blocklog") == expected_key, (
+        f"AFTER: marker ip_blocklog should be today's key {expected_key!r}, got {marker_after.get('ip_blocklog')!r}"
+    )
+
+    # --- Cleanup ---
+    _set_log_rotate_schedule(vm, {"ip_blocklog": "off"})
+    _set_log_reset_keep(vm, {"ip_blocklog": 0})
+    _rm_marker(vm)
+    _rm_log(vm, LOG_IP_BLOCKLOG)
+
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/smoke/test_log_rotate.py]: Applied in `2f9c165` — both new scenarios (5 and 6) now wrap the body in `try` with the schedule/keep/marker/log cleanup in `finally`, so teardown is unconditional. Good catch on the shared module-scoped `deployed_vm` fixture.

Scoped to the two `#416` additions: the four pre-existing log-rotate scenarios (1–4, from ADR-30 #410) use the same inline-cleanup pattern, but they're outside this amendment's diff — retrofitting them to `try/finally` is a reasonable separate consistency pass rather than scope-creep here.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/smoke/test_log_rotate.py]: No need for a separate issue, thanks — it's a trivial consistency pass with no functional risk (the pre-existing scenarios have run fine). It can ride along the next time `test_log_rotate.py` is touched, rather than adding a standalone tracker entry.

---
_Generated by [Claude Code](https://claude.ai/code)_
